# Introduction to Pandas -  Part 1: Data Cleaning

Long gone are the days of pen and paper research. Whether it is for data analysis, data cleaning or data manipulation, proficiency in digital tools has become an unavoidable skill to nurture and exploit. The social sciences are no exception. The reality for a vast number of researchers in psychology today still involves long hours of click-intensive, long and repetitive tasks in front of spreadsheets (*i.e.* Microsoft Excel) or similar statistical analysis software packages (*i.e.* IBM's SPSS). Further, any new iteration of the analysis consists of repeating the same steps and manipulations manually as if starting anew. This not only harms reproducibility (a foundational goal of scientific research and a documented pitfall in the humanities), but also limits the capacity of research teams to develop an expertise in terms of reliable tools and processes that can be reused in different projects and passed on to new students or collaborators. Not to mention, the cumulative hours lost for graduate students performing menial tasks that could be better spent in more creative, cognitive tasks appropriate for their skillsets and interests. 

All of the mentioned limitations can be addressed by even simple programming scripts. For this reason, nowadays graduate students are increasingly encouraged to take on workshops or mini-courses of basic programming and it is not rare to find at least one proficient developer in research teams whose subject of study might very well be far from computer science. In such cases, the most popular choice of programming language is Python. With a very rich ecosystem of data science tools, Python is a general purpose language that allows to create automatized pipelines that can vary from faster, more reliable Excel-like operations, to automatized data gathering through web scraping or web APIs, perform statistical analysis, create beautiful, publication-ready plots, building advanced predictive models and much more.

A cornerstone of this Python data science ecosystem is the pandas library. At its simplest, pandas is the prinicpal data manipulation tool for work on tabular data. That is any type of data that would normally fit into an excel file, a csv file, a sql database, etc., most commonly in the form of text or numerical values. Pandas' main functionality resides in its DataFrame (2D table, as in rows and columns) and Series (1D as in just a column) data structures which contain fast, scalable and easliy customizable functions for all types of tasks, be it data cleaning, manipulation, analysis and even plotting. 

In this first tutorial, we will describe introductory concepts of data processing using Pandas, notably when it comes to the cleaning of a dataset. The concepts learned here should be helpful for anybody that conducted an experiment, a survey or any other data gathering  process that might lead to a somewhat imperfect dataset that needs to be prepared for further analysis. The latter will be covered in a follow-up tutorial.

For teaching purposes, we will be using the Bechdel test dataset. This readily available datset can be easily and freely downloaded from the web. We will explore it soon, but briefly, the dataset is a crowd-sourced list of movies and their Bechdel rating which is a relatively informal measure of the representation of women in movies. Ultimately, we would like to exploit this information to parse trends and other menaingful insights, but given that anybody can contribute to the dataset, sometimes (as often happens in real-life experiments) the data is not properly formatted, elements might be missing and such. The dataset is updated often, so for consitency, for this tutorial, we provide the Bechdel test list as it was available when this manuscript was written. For any studies interested in this data, we encourage downloading the most up to date dataset at the official API.

## Libraries
The first thing to do when working with pandas (and for most Python based projects) is to first import the library itself. Since Pandas doesn't come in the standard python library, you have to make sure that you have installed it in your working environment first. We also import *Path* from pathlib to deal with the paths to different files more clearly.

In [2]:
# Pathing
from pathlib import Path

# Data structures
import pandas as pd

## General Parameters 
It is normally good practice to define any hard-coded values together at the beginning of your code. In our case, we only need the path to the data file and a path to save the data after we are done cleaning it.

In [3]:
data_dir = Path('../data') # define a data folder. The "../" simply means one folder up from where this notebook resides

bechdel_frozen_path = data_dir/'bechdel_frozen.tsv.gz' # the raw frozen data file
bechdel_clean_path = data_dir/'bechdel_clean.tsv.gz' # the path to save the clean data


## The Data - reading and saving data
#### Functions: .read_csv, to.csv

### Reading data
With pandas ready to go, we can now load the data. Often, this might be in the form of a .txt, .csv or .tsv saved on your local machine. Pandas can easily work with excel files, but it requires a few extra steps and special considerations. To keep this introductory tutorial simple, we will skip dealing with excel files. If this is your specific use-case, we recommend for now to simply convert your *.xlsx* file to a *.csv*. A follow-up tutorial exclusively on integrating pandas and excel should be soon become available as well. Also worth noticing is that in cases where the dataset might be too large to fit into memory using pandas would inevitably result into an error. Alternative libraries not covered in this tutorial exist to dela with this type of situations. One particularly worth mentioning is the Dask library. It provides functionality to work with data larger than memory with functionality very similar to pandas (in fact, it uses pandas under the hood and refers to the pandas documentation for many of its functions).

To load the data with pandas, it sufices to use the *.read_csv* method. while the name of it might suggest it only works with *.csv* files, it actually manages to load any similarly formatted file, be it a *.txt*, *.tsv* or other more unique etensions like *.gct* files. As you might have guessed from the path written above, these files can also be compressed (*.gz*) and pandas should have no issue loading it.

By default, *.read_csv* assumes the columns to be separated by commas. Since movie titles themselves might contain commas, the column separator used in the frozen data is a tab (ergo the *.tsv* extension), so we will need to tell the function to use the proper separator ('\t'):

In [4]:
data = pd.read_csv(bechdel_frozen_path, sep='\t')
data

year      imdbid  rating                                   title    id
0     1888    392728.0       0                   Roundhay Garden Scene  8040
1     1892         3.0       0                          Pauvre Pierrot  5433
2     1895    132134.0       0  Execution of Mary, Queen of Scots, The  6200
3     1895        14.0       0           Tables Turned on the Gardener  5444
4     1896       131.0       0                       Une nuit terrible  5406
...    ...         ...     ...                                     ...   ...
8834  2021   5144174.0       3                                Dry, The  9498
8835  2021  10919362.0       3                              Sweetheart  9505
8836  2021  10813940.0       2                       Ginny and Georgia  9501
8837  2021   5109280.0       3                Raya and the Last Dragon  9504
8838  2021   9286908.0       2                             High Ground  9500

[8839 rows x 5 columns]

In [ ]:
data.head().to_latex()

The newly created variable is of the type DataFrame. This is contains not only the data itself as one might find it on a spreadsheet, but also methods to operate on it, some of which will be covered during the tutorial. As with a spreadhseet, the columns are divided into rows and columns. On the top of the dataset, one can see the name of all columns, here the year the movies were realeased, the [imdb](https://www.imdb.com/) id, the bechdel rating, the title of the movies and the internal id of the entry in the website. On the left-hand side is the index of the dataframe. In this case the index is simply the row number, but this is not necessary, and often it is not the case. The index can be any type of row identifier, be it numeric or string and ideally should be unique.

After loading this specific dataset, one thing that should catch every data practitioner's eye is the imdbid column. By default, Pandas will try to guess the dataype of every column when loading with *.read_csv*. For most cases the guess will be accurate. Here, for the imdbid, it wasn't. It actually interpreted it as a float type. Not only, it shouldn't be interpreted as a float, but any numeric type would reulst in innapropriate conversionons because any entry that starts with 0s would lose them. For example, the id for Pauvre Pierrot (entry 1) which has an imdbid of 0000003 is converted to 3.0.

Luckily among the options one can pass to *.read_csv*, it is possible to specify the data type to be used for each column. We do so as follows:

In [7]:
data = pd.read_csv(bechdel_frozen_path, sep='\t', dtype={'imdbid':str})
data

year    imdbid  rating                                   title    id
0     1888   0392728       0                   Roundhay Garden Scene  8040
1     1892   0000003       0                          Pauvre Pierrot  5433
2     1895   0132134       0  Execution of Mary, Queen of Scots, The  6200
3     1895   0000014       0           Tables Turned on the Gardener  5444
4     1896   0000131       0                       Une nuit terrible  5406
...    ...       ...     ...                                     ...   ...
8834  2021   5144174       3                                Dry, The  9498
8835  2021  10919362       3                              Sweetheart  9505
8836  2021  10813940       2                       Ginny and Georgia  9501
8837  2021   5109280       3                Raya and the Last Dragon  9504
8838  2021   9286908       2                             High Ground  9500

[8839 rows x 5 columns]

Now we see the imdb id properly formatted. Generally, if one knows in advance the data types and its not incredibly troublesome to do so, one could/should specify the data types of the columns in the *read_csv* function itself with the keyword *dtypes*.

Other important keywords for *.read_csv* that can be found on the [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) worth remembering are:

* *header*,  *index_col*: To specify the row to use as headers and/or the row to use as index. You'll mostly use it when you don't actually have a header and pandas still ueses your first row of data  as a header and/or when you have an index but pandas think it's another data column, and adds a secodn index to the dataframe.

* *usecols*: if you only need some columns, but not all, you can specify which ones here.

* *names*: you can specify the name of the columns if you didn't specify the header.

* *delim_whitespace*: if the separator is blank space between columns (tab '\t' included) you can use this to read them.

There's quite a lot more of options available to users. If at any point it is needed to load a dataframe in some special way not covered by these commands, it is a good idea to look at the documentation first. Chances are that the special use-case is already covered by an existing option.

We haven't done any manipulations yet, but before doing so, it is good practice (somewhat unavoidable and mandatory) to save a copy of the raw data in another file to avoid risking performing irreversible changes that might ruin it, and with it, any chances of a succesful, reproductible project.

### Saving data
To save the data locally, we use a very similar function to the one used to read the data. For now, we can use the default parameters:

In [22]:
data.to_csv(bechdel_clean_path)

That's the most basic way to save the file. At the path defined above, a new 'bechdel_clean.tsv.gz' file should have appeared. If one were to open it, you could see a neat table with all of the data in it separated by commas, not tabs as we had initially. Aslo, depending on your version of pandas, even though the file's extension is '.gz', it might not be compressed. Also, you might see that a new column was added, corresponding to the index number of the data. Since this information is relatively redundant (it is simply the line number), we might want to remove it

We can fix these short comings With the following options from the [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html):
* *sep*: if you pan on using a column delimiter other than the comma ',' (useful if you will be storing text with commas in them as in here).
    * Notably to save with a tab delimiter (a .tsv essentialy), simply use sep='\t' (of course, you would need to change the extension name from the filename yourself).
* *header* and *index*: to determine whether you want to save the index column and/or the header row.
* *compression*: if you want your file to be compressed when saved (great for large files). For example, using 'gzip' (it is good practice to be explicit in the name of the file and add the compression to the extension), will force pandas to compress the file irrespective of the name of the file provided.

The modified command looks like this

In [23]:
data.to_csv(bechdel_clean_path, sep='\t', index=False, compression='gzip')

This covers the reading and writing of pandas dataframe. In the next section, we will start the actual process of gaining a basic understanding of the data and cleainng it

## The DataFrame - Understanding the data container
#### Functions: .head, .tail, .sample, .info, .shape, .columns, .dtypes, sort_values

Before looking at the details of the data, it pays off to have an idea of what the dataframe itself looks like: how big is it? What type of data is encoded? Having a good understanding of these fundamental properties will help guide and inspiring anything done to the content of the dataset iteslf.

First, and most common, its to look at some entries to get a glimpse of the data. To look at the top, the bottom or some randomly chosen sample of rows, one uses respectively:

In [24]:
data.head(3)

year   imdbid  rating                                   title    id
0  1888  0392728       0                   Roundhay Garden Scene  8040
1  1892  0000003       0                          Pauvre Pierrot  5433
2  1895  0132134       0  Execution of Mary, Queen of Scots, The  6200

In [25]:
data.tail(2)

year   imdbid  rating                     title    id
8837  2021  5109280       3  Raya and the Last Dragon  9504
8838  2021  9286908       2               High Ground  9500

In [26]:
data.sample(6)

year   imdbid  rating                              title    id
6604  2013  1951264       3   Hunger Games: Catching Fire, The  4851
5224  2009  1233219       2  My Son, My Son, What Have Ye Done  3165
5970  2011  1411238       3                No Strings Attached  2275
3762  2002  0303678       1              Armitage: Dual Matrix  4429
426   1942  0034492       1                              Bambi  1116
5342  2009  0758746       2                    Friday the 13th  4570

All three functions accept the number of wanted entreis as input. The head and tail functions By default, head and tail return 5 elements. Sample by default will return only one element. To tailor the sampling itself, the function gives access to parameters such as sampling weights, a random seed and the possibility to sample columns if needed. As always, it is recommended to look at the [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html), when first using any function.

For a general inspection of the dataframe properties, one can use the *.info* function:

In [27]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8839 entries, 0 to 8838
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   year    8839 non-null   int64 
 1   imdbid  8835 non-null   object
 2   rating  8839 non-null   int64 
 3   title   8839 non-null   object
 4   id      8839 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 345.4+ KB


This returns the number of entries, and information on the columns: the name of the column, the number of non-null/na/missing elements and the data type of the column. One could get the same information using the following functions (which often are useful in many different contexts):

In [28]:
data.shape

(8839, 5)

In [29]:
data.columns

Index(['year', 'imdbid', 'rating', 'title', 'id'], dtype='object')

In [30]:
data.dtypes

year       int64
imdbid    object
rating     int64
title     object
id         int64
dtype: object

In [31]:
data.notna().sum()

year      8839
imdbid    8835
rating    8839
title     8839
id        8839
dtype: int64

While the first three are relatively self-explanatory, the fourth one requires more attention. The *.notna()* is discussed in the next section and in a nutshell tells whether datapoints are not na (not missing) or na (missing). Since True and False are respectively equivalent to 1 and 0, by summing them (*.sum* by default sums across rows), one can calculate how many nans are in each column.

More importantly though,  that command shows an important feature of pandas: one can chain a series of functions to obtain more complex results than a single function my provide, without having to store the results in intermediate data structures.

We can see this sort of chainng in another example. If one was interested in looking at the five oldest (in terms of release year) movies in the dataset, one would first need to sort them by year (using the *.sort_values* function) and then respectively the *.head* function described above: 

In [34]:
data.sort_values('year').head()

year   imdbid  rating                                   title    id
0  1888  0392728       0                   Roundhay Garden Scene  8040
1  1892  0000003       0                          Pauvre Pierrot  5433
2  1895  0132134       0  Execution of Mary, Queen of Scots, The  6200
3  1895  0000014       0           Tables Turned on the Gardener  5444
4  1896  0000131       0                       Une nuit terrible  5406

## The Data - Clean it before using it to save future headaches
#### .isna, .notna

It is always very tempting to begin, manipulating, analysing and plotting the data right away. However, unless one can be sure it has been done before, it is important first to perform some basic data quality checks and cleaning. At the very least, and most commonly, that should involve dealing with NaN values. 

Irrespective of how well the experiment or survey designed for a research project is, participants might fail to fully answer questions or might drop out before the end of the study, measuring devices might fail to record at all times or any number of events could result on the data being incomplete. If left unchecked, at best, these Nan/missing values will result in explicit errors when running analysis software that will stop their execution. At worst, the missing values will go unnoticed and cause biased results leading to false conclusions.

Since it is such a common use-case, pandas comes well equipped to deal with missing values. First, one can look if they are found in all columns or if there is a specifc vulnurable one. For that, one uses the opposite of the *.notna* function used before, the *.isna* function:

In [35]:
data.isna().sum()

year      0
imdbid    4
rating    0
title     0
id        0
dtype: int64

So, there is only one column with missing value. To find which rows within that column it is first necessary to introduce methods for row indexing

### Short aside - Indexing
#### Functions: .loc, .iloc
The first ype of indexing introduced here is boolean indexing, that is, using an array (with the same length as the dataframe) of True's and False's that determine wheter a row should be kept or not. For that, the now familiar *.isna* function will be usfeul. First, since it has previously been used right before a *.sum*, it is useful to see what it looks like by itself:

In [36]:
data.isna()

year  imdbid  rating  title     id
0     False   False   False  False  False
1     False   False   False  False  False
2     False   False   False  False  False
3     False   False   False  False  False
4     False   False   False  False  False
...     ...     ...     ...    ...    ...
8834  False   False   False  False  False
8835  False   False   False  False  False
8836  False   False   False  False  False
8837  False   False   False  False  False
8838  False   False   False  False  False

[8839 rows x 5 columns]

As explained before, for every element, it returns True if the value is a NaN (if the data is missing for example) and False if the value is not a NaN (if it is a valid entry). Since only the 'imdbid' column is the only one with NaNs, it is more helpful to only use *.isna* on it. There are multiple ways of selecting a single column:

In [41]:
data['imdbid']

0        0392728
1        0000003
2        0132134
3        0000014
4        0000131
          ...   
8834     5144174
8835    10919362
8836    10813940
8837     5109280
8838     9286908
Name: imdbid, Length: 8839, dtype: object

In [47]:
data.imdbid

0        0392728
1        0000003
2        0132134
3        0000014
4        0000131
          ...   
8834     5144174
8835    10919362
8836    10813940
8837     5109280
8838     9286908
Name: imdbid, Length: 8839, dtype: object

These two are interchangealbe in most cases and might be more a question of personal taste. One exception is if the name of the column contains spaces in it. For example, if the column had been named 'imdb id', then only data\['imdb id'\] could have been used without running into python syntax problems.

In both cases, the resulting object is a pandas Series, the one-dimensional equivalent of a dataframe. While sometimes the difference might not be important, Series and DataFrames are different data structures, and functions available in one might not be available in the other. Therefore, if one were to subset a single column and preferred that the resulting column data structure was a DataFrame, one can trick pandas by selecting the column as a list with one element:

In [60]:
data[['imdbid']]

imdbid
0      0392728
1      0000003
2      0132134
3      0000014
4      0000131
...        ...
8834   5144174
8835  10919362
8836  10813940
8837   5109280
8838   9286908

[8839 rows x 1 columns]

This is a bit of a trick, because passing a list of column names is a way to select multiple columns

In [79]:
data[['year', 'imdbid']].head(3)

year   imdbid
0  1888  0392728
1  1892  0000003
2  1895  0132134

With that in mind, it is possible to generate the boolean array for indexing rows, by using *.isna* only on the 'imdbid' column (as a 1D aray, in this case a pandas Series, but it could very well be a pyton boolean array, tuple or other similar data structures).

In [65]:
data.imdbid.isna()

0       False
1       False
2       False
3       False
4       False
        ...  
8834    False
8835    False
8836    False
8837    False
8838    False
Name: imdbid, Length: 8839, dtype: bool

While one could select columns with square brackets, to select rows, it is necessary to use the more general *.loc* (as in *loc*ation) to select the specifc rows: 

In [66]:
data.loc[data.imdbid.isna(), ['year', 'imdbid']]

year imdbid
7602  2015    NaN
8208  2017    NaN
8280  2017    NaN
8545  2019    NaN

The *.loc* method expects two elements, separated by a comma: the rows to subset, and the columns to subset. For the rows, it sufficed to use the boolean array that selects with a True those elements one wants to keep. For the columns, as before, one can select a subset of columns by passing them them as a list. If instead, all columns were needed, the naive and wrong way to do it could be to pass the list of all columns to *.loc*, which can be done practically as:

In [68]:
data.loc[data.imdbid.isna(), data.columns]

year imdbid  rating                   title    id
7602  2015    NaN       3       Danish Girl , The  9081
8208  2017    NaN       3            Wonder Woman  9294
8280  2017    NaN       3            Wonder Woman  9293
8545  2019    NaN       3  Rise Of Skywalker, The  9098

A more practical way is to use the more traditional python indexing operator ":". For clarity, first, we can use it in the cases where one needs to retrieve all columns after 'imdbid', all columns before 'title' or all columns between them:

In [82]:
data.loc[data.imdbid.isna(), 'imdbid':]

imdbid  rating                   title    id
7602    NaN       3       Danish Girl , The  9081
8208    NaN       3            Wonder Woman  9294
8280    NaN       3            Wonder Woman  9293
8545    NaN       3  Rise Of Skywalker, The  9098

In [84]:
data.loc[data.imdbid.isna(), 'imdbid':'title']

imdbid  rating                   title
7602    NaN       3       Danish Girl , The
8208    NaN       3            Wonder Woman
8280    NaN       3            Wonder Woman
8545    NaN       3  Rise Of Skywalker, The

In [85]:
data.loc[data.imdbid.isna(), :'title']

year imdbid  rating                   title
7602  2015    NaN       3       Danish Girl , The
8208  2017    NaN       3            Wonder Woman
8280  2017    NaN       3            Wonder Woman
8545  2019    NaN       3  Rise Of Skywalker, The

To select all columns then, we can simply do

In [69]:
data.loc[data.imdbid.isna(), :]

year imdbid  rating                   title    id
7602  2015    NaN       3       Danish Girl , The  9081
8208  2017    NaN       3            Wonder Woman  9294
8280  2017    NaN       3            Wonder Woman  9293
8545  2019    NaN       3  Rise Of Skywalker, The  9098

Which essentially means columns without bounding on the left, and withoung bounding on the right (so all of them). The same can also be done for rows:

In [9]:
data.loc[8835:, :]

year    imdbid  rating                     title    id
8835  2021  10919362       3                Sweetheart  9505
8836  2021  10813940       2         Ginny and Georgia  9501
8837  2021   5109280       3  Raya and the Last Dragon  9504
8838  2021   9286908       2               High Ground  9500

In [10]:
data.loc[:4, :]

year   imdbid  rating                                   title    id
0  1888  0392728       0                   Roundhay Garden Scene  8040
1  1892  0000003       0                          Pauvre Pierrot  5433
2  1895  0132134       0  Execution of Mary, Queen of Scots, The  6200
3  1895  0000014       0           Tables Turned on the Gardener  5444
4  1896  0000131       0                       Une nuit terrible  5406

In [86]:
data.loc[:, ['year', 'imdbid']]

year    imdbid
0     1888   0392728
1     1892   0000003
2     1895   0132134
3     1895   0000014
4     1896   0000131
...    ...       ...
8834  2021   5144174
8835  2021  10919362
8836  2021  10813940
8837  2021   5109280
8838  2021   9286908

[8839 rows x 2 columns]

As a recap after the *.loc* tangent, one can index using an array of booleans, in this case a pandas Series created by *.isna*, to indivudally select which elemnts to keep and which to leave.

A second method of indexing is by only passing the index of the rows to keep. Those can be found from the previous result with the equivalent of *.columns* for rows:

In [21]:
nan_idx = data.loc[data.imdbid.isna(), :].index
nan_idx

Int64Index([7602, 8208, 8280, 8545], dtype='int64')

Even though a small subset of indeces is used instead of a long boolean array, the *.loc* is still the method to use for indexing:

In [87]:
data.loc[nan_idx, :]

year imdbid  rating                   title    id
7602  2015    NaN       3       Danish Girl , The  9081
8208  2017    NaN       3            Wonder Woman  9294
8280  2017    NaN       3            Wonder Woman  9293
8545  2019    NaN       3  Rise Of Skywalker, The  9098

*.loc[ ]* is great to get data from the dataframe if you have a list of indeces or columns you want to access or some type of boolean array like the one we get from *.isna()*. If instead one wanted to use the specific numeric position (*e.g.* the 50th to 60th row, 2rd to 4th column), then  *.iloc[]* must be used instead:

In [11]:
data.iloc[50:63:3, 2:5]

rating                                 title    id
50       0                      Good Glue Sticks  5615
53       1           Le barometre de la fidelite  6211
56       0                        A Trip to Mars  5685
59       2                             Cleopatra  2003
62       0  Voyage of the Bourrichon Family, The  5364

In this case, the distinction of loc and iloc might not be particularly clear when it comes to indexing rows since the indeces themselves are the row number. As explained earlier, this is not necessarily true in every DataFrame. As a quick example, one could make the id column the new index:

In [17]:
data.set_index('id')

year    imdbid  rating                                   title
id                                                                  
8040  1888   0392728       0                   Roundhay Garden Scene
5433  1892   0000003       0                          Pauvre Pierrot
6200  1895   0132134       0  Execution of Mary, Queen of Scots, The
5444  1895   0000014       0           Tables Turned on the Gardener
5406  1896   0000131       0                       Une nuit terrible
...    ...       ...     ...                                     ...
9498  2021   5144174       3                                Dry, The
9505  2021  10919362       3                              Sweetheart
9501  2021  10813940       2                       Ginny and Georgia
9504  2021   5109280       3                Raya and the Last Dragon
9500  2021   9286908       2                             High Ground

[8839 rows x 4 columns]

Accesing the first three entries with *.loc* and *.iloc* then becomes very different:

In [18]:
data.set_index('id').loc[:6200, :]

year   imdbid  rating                                   title
id                                                                 
8040  1888  0392728       0                   Roundhay Garden Scene
5433  1892  0000003       0                          Pauvre Pierrot
6200  1895  0132134       0  Execution of Mary, Queen of Scots, The

In [19]:
data.set_index('id').iloc[:3, :]

year   imdbid  rating                                   title
id                                                                 
8040  1888  0392728       0                   Roundhay Garden Scene
5433  1892  0000003       0                          Pauvre Pierrot
6200  1895  0132134       0  Execution of Mary, Queen of Scots, The

### Dealing with NaNs
#### .dropna, .fillna, .rolling, astype, .set_index
Figuring out where the NaN values was simply the first step and it provided an excellenmt opportunity to also discuss accessing data in pandas. As a reminder, this is the what the missing data looks like in this dataset:

In [22]:
data.loc[nan_idx, :]

year imdbid  rating                   title    id
7602  2015    NaN       3       Danish Girl , The  9081
8208  2017    NaN       3            Wonder Woman  9294
8280  2017    NaN       3            Wonder Woman  9293
8545  2019    NaN       3  Rise Of Skywalker, The  9098

Dealing with missing data is normally a complicated subject whose solution can be highly dependent on the data source itself and methodological considerations of the study and analysis at play.

The first and simplest way to deal with missing data if one can afford it is to drop the row or column containint NaNs In pandas this is achieved with *.dropna*:

In [23]:
data.dropna()

year    imdbid  rating                                   title    id
0     1888   0392728       0                   Roundhay Garden Scene  8040
1     1892   0000003       0                          Pauvre Pierrot  5433
2     1895   0132134       0  Execution of Mary, Queen of Scots, The  6200
3     1895   0000014       0           Tables Turned on the Gardener  5444
4     1896   0000131       0                       Une nuit terrible  5406
...    ...       ...     ...                                     ...   ...
8834  2021   5144174       3                                Dry, The  9498
8835  2021  10919362       3                              Sweetheart  9505
8836  2021  10813940       2                       Ginny and Georgia  9501
8837  2021   5109280       3                Raya and the Last Dragon  9504
8838  2021   9286908       2                             High Ground  9500

[8835 rows x 5 columns]

The dataset went from having 8839 rows to 8835, which corresponds to the number of nans originally in the dataset. TO confirm, one can use the same method as before:

In [24]:
data.dropna().isna().sum()

year      0
imdbid    0
rating    0
title     0
id        0
dtype: int64

As with everything that has been done up to this point, in the way *.dropna* was used, the changes were not actually applied to the dataframe permanently

In [26]:
print(data.shape)
data.isna().sum()

(8839, 5)


year      0
imdbid    4
rating    0
title     0
id        0
dtype: int64

All the orignal rows are still there. To store any changes, including the new NaN-less data, it is necessary to save it as a new variable or overwrite the currently used dataframe (the one in memory now, not the saved file):

clean_data = data.dropna()

data = data.dropna()

Another option is to use the **inplace** keyword. *inplace* apears in several pandas functions and essentially overwrites the dataframe without actually having to use syntax like data = data.dropna():

data.dropna(inplace=True)

The use of *.inplace* though is generally not recommended and it is adviced against for beginners. Instead, creating intermediary DataFrames or overwriting the original one should be used.

For this particular dataset, there are better ways of dealing with NaNs than dropping rows, so *.dropna* will not be necessary.

The second option when dealing with NaNs, when it is preferred no to drop any data is to fill or impute the missing values. A  helpful function for it is the aptly named *.fillna()*. *.fillna* offers some great functionality to solve most cases: It is possible to pass a single value which will replace all Nans for a single or for multiple columns:

In [36]:
data.fillna(-99).loc[nan_idx, :]

year imdbid  rating                   title    id
7602  2015    -99       3       Danish Girl , The  9081
8208  2017    -99       3            Wonder Woman  9294
8280  2017    -99       3            Wonder Woman  9293
8545  2019    -99       3  Rise Of Skywalker, The  9098

In [37]:
data.fillna({'imdbid':'a new value', 'rating':'no nans on this column'}).loc[nan_idx, :]

year       imdbid  rating                   title    id
7602  2015  a new value       3       Danish Girl , The  9081
8208  2017  a new value       3            Wonder Woman  9294
8280  2017  a new value       3            Wonder Woman  9293
8545  2019  a new value       3  Rise Of Skywalker, The  9098

A common choice in this type of cases is to fill every column by its mean. Since here the column containg NaNs is actually a string, for demonstration purposes, it first needs to be converted to a float temporarily:

In [49]:
data.fillna(data.astype({'imdbid':float}).mean()).loc[nan_idx, :]

year          imdbid  rating                   title    id
7602  2015  1365556.846633       3       Danish Girl , The  9081
8208  2017  1365556.846633       3            Wonder Woman  9294
8280  2017  1365556.846633       3            Wonder Woman  9293
8545  2019  1365556.846633       3  Rise Of Skywalker, The  9098

*.fillna* also offers the choice of a few predefined methods. 'ffill' (forward-fill) can be used to repeat the last know measure and 'bfill' (backwards-fill) to use the next known measure to replace NaNs:

In [51]:
data.fillna(method='ffill').loc[nan_idx, :]

year   imdbid  rating                   title    id
7602  2015  4428814       3       Danish Girl , The  9081
8208  2017  5155780       3            Wonder Woman  9294
8280  2017  7341676       3            Wonder Woman  9293
8545  2019  5363618       3  Rise Of Skywalker, The  9098

This can be compared to the values right before the rows containitng missing values:

In [55]:
data.loc[nan_idx - 1, :]

year   imdbid  rating                    title    id
7601  2015  4428814       1  La loi du march&eacute;  6279
8207  2017  5155780       3           Discovery, The  7581
8279  2017  7341676       3                    OM+ME  7895
8544  2019  5363618       1           Sound of Metal  9448

Those are some basic ways of dealing with NaNs. As mentioned before, there exist many ways of dealing with NaNs and each diffrent dataset might require a different method. If more complex methods are necessary though, those presented before can be readily extende. For example one might need to replace the NaNs with a rolling average (averaging over the nearest data points as opposed to using the whole column). That can be done as follows:

In [56]:
data['imdbid'].rolling(window=6, min_periods=1).mean()

0       3.927280e+05
1       1.963655e+05
2       1.749550e+05
3       1.312198e+05
4       1.050020e+05
            ...     
8834    7.956427e+06
8835    9.595310e+06
8836    9.883262e+06
8837    8.962566e+06
8838    8.601042e+06
Name: imdbid, Length: 8839, dtype: float64

This essentially asked to use a sliding window of size 6 (the number of elements that are considered at one time) and the minimum number of elements required to use (in case many consecutive NaNs are present). On the rolling operation one can attach any aggregating function, in this case the mean. Since this is potentially a more advanced use-case, it is highly recommended to visit the [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rolling.html) to learn more about the rolling function.

The previous method simply calculated the rolling average. If this wanted to be applied to the dataset, it can be assigned to the column itself with

data['imdbid'] = data['imdbid'].rolling(window=6, min_periods=1).mean()

In [68]:
data['imdbid'].rolling(window=6, min_periods=1).mean()[nan_idx]

7602    2846128.4
8208    4582077.2
8280    5601607.6
8545    7144776.8
Name: imdbid, dtype: float64

Since one can't really calculate a given movie's imdb-id from other movies imdb-id, in this case, to replace the NaNs it is necessary to find them. Luckily there aren't many too do so manually:

In [69]:
data.loc[nan_idx, :]

year imdbid  rating                   title    id
7602  2015    NaN       3       Danish Girl , The  9081
8208  2017    NaN       3            Wonder Woman  9294
8280  2017    NaN       3            Wonder Woman  9293
8545  2019    NaN       3  Rise Of Skywalker, The  9098

Before doing so, it is worth noticing that Wonder Woman (2017) appears on the list twice. This reveals the need to first discuss the second most common data cleaning process: dealing with duplicattes.

###  Dealing with Duplicates
#### .duplicated, .drop, .reset_index
First, it is good to know if there are any more duplicates. We can do this similarly to how NaNs were found, by using the appropriate boolean indexing. Here, this is achieved using *.duplicated*:

In [70]:
data[data.duplicated(subset=['year', 'title'], keep=False)]

year     imdbid  rating                   title    id
182   1931    0021814       2                 Dracula  1985
192   1931    0021815       3                 Dracula  8213
839   1959    0053285       3         Sleeping Beauty  9209
848   1959      53285       3         Sleeping Beauty   474
1783  1983    0086425       3     Terms of Endearment  4449
1792  1983    0086425       1     Terms of Endearment  4448
2915  1997    0117056       3                   Ayneh  4380
2998  1997    0117056       3                   Ayneh  4381
6010  2011    2043900       3  Last Call at the Oasis  4889
6092  2011    2043900       3  Last Call at the Oasis  4907
7068  2014    2180411       3          Into the Woods  6000
7238  2014  002180411       3          Into the Woods  8702
8182  2017    0451279       3            Wonder Woman  7241
8208  2017        NaN       3            Wonder Woman  9294
8280  2017        NaN       3            Wonder Woman  9293

As with isna(), the *.duplicated* function returns a Series of booleans specifying if the row has a duplicate or not. The *subset* parameter allows to consider a group of columns for repeats. In this case, it was important to consider a duplicate anything that shares the same title and year. That way, anything like a reboot or movies with the same title wouldn't be wrongly flagged as a duplicate. The *keep* parameter is used to determine which duplicate gets assigned a True. With False, it means to flag all duplicate copies. Alternatively, one can use the 'first' or 'last' option if only the first or last occurence of the duplicates should be considered respectively.
 
Again, as with NaNs, if one didn't care much about the identiy of the duplicates, they can be quickly removed with the aptly named function *drop_duplicates*.

In [71]:
data.drop_duplicates(subset=['year', 'title'], keep='last')

year    imdbid  rating                                   title    id
0     1888   0392728       0                   Roundhay Garden Scene  8040
1     1892   0000003       0                          Pauvre Pierrot  5433
2     1895   0132134       0  Execution of Mary, Queen of Scots, The  6200
3     1895   0000014       0           Tables Turned on the Gardener  5444
4     1896   0000131       0                       Une nuit terrible  5406
...    ...       ...     ...                                     ...   ...
8834  2021   5144174       3                                Dry, The  9498
8835  2021  10919362       3                              Sweetheart  9505
8836  2021  10813940       2                       Ginny and Georgia  9501
8837  2021   5109280       3                Raya and the Last Dragon  9504
8838  2021   9286908       2                             High Ground  9500

[8831 rows x 5 columns]

The options used are not very different from *.dropna* before and *.duplicated* right above. It was essentially tasked with considering only the year and title columns (instead of the whole row) and in this case, the keep keyword was used todrop all but the last of all the duplicate sets. Again, to save the resultsone should use the *inplace* parameter (not recommende) or reassign with data = data.drop_duplicates(...).

by close inspection of the duplicates though, for this given dataset, it is more useful to manually inspect all copies of the duplicated entries (there are not hat many), in order to determine which ones to keep. In many applications, this is an important consideration.

In [72]:
data[data.duplicated(subset=['year', 'title'], keep=False)]

year     imdbid  rating                   title    id
182   1931    0021814       2                 Dracula  1985
192   1931    0021815       3                 Dracula  8213
839   1959    0053285       3         Sleeping Beauty  9209
848   1959      53285       3         Sleeping Beauty   474
1783  1983    0086425       3     Terms of Endearment  4449
1792  1983    0086425       1     Terms of Endearment  4448
2915  1997    0117056       3                   Ayneh  4380
2998  1997    0117056       3                   Ayneh  4381
6010  2011    2043900       3  Last Call at the Oasis  4889
6092  2011    2043900       3  Last Call at the Oasis  4907
7068  2014    2180411       3          Into the Woods  6000
7238  2014  002180411       3          Into the Woods  8702
8182  2017    0451279       3            Wonder Woman  7241
8208  2017        NaN       3            Wonder Woman  9294
8280  2017        NaN       3            Wonder Woman  9293

Since the specifc of the dataset are less significant for the tutorial, here is a quick summary of how duplicates were dealt with:
1. The Dracula (1931) movie entries don't share their imdb ids.
    * Are they two different Dracula movies from 1931? Not likely, need to check the imdb id manually even if not a NaN.
2. One of the Sleeping Beauty (1959) imdb ids is missing a couple of zeros (or the other has extra zeros)
    * Check id manually to decide which one to keep
3. Terms of Endearment (1983) shares the ids but not the bechdel rating
    * Watching the movie and deciding the rating would be the best thing to do. A quick wiki search though shows the movie covers 30 years of the relationship between a mother and a daughter so a rating of 3 is more plausible.
4. Ayneh (1997) and Last Call at the Oasis (2011) seem to simply be differnt ids in the webiste's own database, so true duplicates here
    * Drop any of the two
5. Into the Woods (2014) also missing (or extra) zeros on the imdbid
    * Check id manually to decide which one to keep
6. Wonder Woman (2017) has 3 entries, two with NaNs for the imdbid and one with an id
    * Worth checking the imdb id present and drop the NaN rows
    
Perfoming all of these manual checks, the following rows (by row index). Be mindful that for the first itme, the results of the function are used to overwrite the dataset. If the next block of code is run more than once it will throw an error because the rows would already have been dropped by the second run.

In [73]:
to_drop = [192, 848, 1792, 2998, 6010, 7238, 8208, 8280]
data = data.drop(to_drop)

Here, the *.drop* function was used to drop rows by index. If instead, a whole column needed to be dropped, one can pass its name (or a lit of names) and specify that it is a column and not a row, by adding *axis=1* to the inputs.

Since some rows were droped and the index in this case is not informative (it is simply the row number), it would be good practice to also reset the index, so it doesn't skip any numbers (those corresponding to the dropped duplicates). That can be done with *reset_index*. By default, that would look like this:

In [75]:
data.reset_index()

index  year    imdbid  rating                                   title  \
0         0  1888   0392728       0                   Roundhay Garden Scene   
1         1  1892   0000003       0                          Pauvre Pierrot   
2         2  1895   0132134       0  Execution of Mary, Queen of Scots, The   
3         3  1895   0000014       0           Tables Turned on the Gardener   
4         4  1896   0000131       0                       Une nuit terrible   
...     ...   ...       ...     ...                                     ...   
8826   8826  2021   5144174       3                                Dry, The   
8827   8827  2021  10919362       3                              Sweetheart   
8828   8828  2021  10813940       2                       Ginny and Georgia   
8829   8829  2021   5109280       3                Raya and the Last Dragon   
8830   8830  2021   9286908       2                             High Ground   

        id  
0     8040  
1     5433  
2     6200  
3     5444  
4     5406  
...    ...  
8826  9498  
8827  9505  
8828  9501  
8829  9504  
8830  9500  

[8831 rows x 6 columns]

That simply pushed the index inside the dataframe as a brand new column and then created a new index. If one prefers completely dropping the index without adding an extra column, the drop keyword can be used:

In [76]:
data = data.reset_index(drop=True)
data

year    imdbid  rating                                   title    id
0     1888   0392728       0                   Roundhay Garden Scene  8040
1     1892   0000003       0                          Pauvre Pierrot  5433
2     1895   0132134       0  Execution of Mary, Queen of Scots, The  6200
3     1895   0000014       0           Tables Turned on the Gardener  5444
4     1896   0000131       0                       Une nuit terrible  5406
...    ...       ...     ...                                     ...   ...
8826  2021   5144174       3                                Dry, The  9498
8827  2021  10919362       3                              Sweetheart  9505
8828  2021  10813940       2                       Ginny and Georgia  9501
8829  2021   5109280       3                Raya and the Last Dragon  9504
8830  2021   9286908       2                             High Ground  9500

[8831 rows x 5 columns]

### Aside - Finish filling NaNs
Now that the dataset is free of duplicates, it is possible to finally finish dealing with NaN values. In this case, a quick search of the two movies on imdbid is the best solution:

In [77]:
nan_idx = data.loc[data.imdbid.isna()].index
data.loc[nan_idx]

year imdbid  rating                   title    id
7596  2015    NaN       3       Danish Girl , The  9081
8537  2019    NaN       3  Rise Of Skywalker, The  9098

A similar syntax as the one used before can be used here to fill the NaN values.

In [79]:
missing_ids = {7596:'0810819', 8537:'2527338'}
data.imdbid = data.imdbid.fillna(missing_ids)
data.loc[nan_idx, :]

year   imdbid  rating                   title    id
7596  2015  0810819       3       Danish Girl , The  9081
8537  2019  2527338       3  Rise Of Skywalker, The  9098

That concludes two of the main data cleaning tasks encountered when first encountering a dataset. Before finishing though, there is another functionality of pandas worth mentioning

##  String columns - Processing badly written text
#### .str, .str.replace

Through out the tutorial, another noticeable issue on the data quality might have become apparent. In the titles column, a seemingly strange series of characters (\&#39;) had taken the place of apostrophes ('). While this is due to different character encodings and not necessarily the data gathering or storing, errors in inputted texts are a common occurence in research that deals with patients/subjects filling questionaires, researchers taking notes, etc. This can even be quite inconsequential for a person, such as the answer to a 'yes' or 'no' question being 'yeah'. However, when running analysis, consistency in the data is essential for computers, so knowing how to efficiently clean string data is an important tool.

Within pandas, one approach is to use the functionality of *.str*. To find all the rows in the title column that have the wrong encoding for apostrophes, one can use the function *.str.contains*:

In [81]:
seq = '&#39;'
data.title.str.contains(seq)

0       False
1       False
2       False
3       False
4       False
        ...  
8826    False
8827    False
8828    False
8829    False
8830    False
Name: title, Length: 8831, dtype: bool

As before, *.contains*, returns a boolean array which indicates whether the specific entry contains the string sequence seq. This can be used to index the dataframe itself

In [84]:
bol_idx = data.title.str.contains(seq)
data.loc[bol_idx, ['title']]

title
11    Astronomer&#39;s Dream; or, The Man in the Moo...
15                       Hamlet ( Le Duel d&#39;Hamlet)
24                          Grandma&#39;s Reading Glass
26                  L&#39;homme a la tete en caoutchouc
29                  Mephistopheles&#39; School of Magic
...                                                 ...
8704                     Where&#39;d You Go, Bernadette
8744       Mariah Carey&#39;s Magical Christmas Special
8775                                 I&#39;m Your Woman
8776                       Ma Rainey&#39;s Black Bottom
8825                                 Finding &#39;Ohana

[392 rows x 1 columns]

Using the boolean index together with *.loc* and the trick to subset only one column as a dataframe, it was possible to subset only those titles that contain the improperly encoded apostrophes. In this case, there are 392 titles to correct, which would be tedious and wasteful to correct manually, one at a time.

For this purpose, it is far more efficient to use *.str.replace*:

In [86]:
data.title = data.title.str.replace(seq, "'")
data.loc[bol_idx, ['title']]

title
11    Astronomer's Dream; or, The Man in the Moon, The
15                          Hamlet ( Le Duel d'Hamlet)
24                             Grandma's Reading Glass
26                     L'homme a la tete en caoutchouc
29                     Mephistopheles' School of Magic
...                                                ...
8704                        Where'd You Go, Bernadette
8744          Mariah Carey's Magical Christmas Special
8775                                    I'm Your Woman
8776                          Ma Rainey's Black Bottom
8825                                    Finding 'Ohana

[392 rows x 1 columns]

As one can guess, the *.str.replace* function take the sequence to be replaced and the sequence with which to replace it (&#39; and ' respectively) and performs the change in all rows. While that might be all the string cleaning required for this dataset, *.str* is equipped with a plethora of other useful functions that should address most, if not all string operations that might be needed when cleaning a dataset. Again, it is strongly adviced to refer to the official documentation of *.str* whenever string operations on pandas columns are needed.

That should conclude the data cleaning process for the dataset. Before doing anything else, it is important to save it as shown above:

In [87]:
data.to_csv(bechdel_clean_path, index=False, sep='\t', compression='gzip')

## Conclusion
While far from comprehensive, this tutorial showed how to perform basic data cleaning operations on a realistic dataset, analogous to what a researcher might encounter in their own projects. It focused on reading and saving files, understanding the DataFrame object and dealing with duplicate entries, missing data as well as inproper string columns. These concepts also allowed to introduce other core principiles of data manipulation in pandas such as indexing and dropping unwanted rows and columns.

This should serve as a starting point for further exploration and exploitation of pandas functionality and provide an extra layer of consistency, efficency and transparency to research projects in the humanities and psychology in particular. In a future tutorial, we will build upon the concepts taught in this work, to provide a similar baseline into exploratory data analysis and basic modelling which should provide the complementary pandas tools to succesfully take a research question from raw data to publication ready plots of new and interesting insights.